Implementing filtering to new contributor admin dashboard: GSoCM2.4

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #18907 and subpart of #17440 .
2. **FYI: This PR is the branch of #18946 so some changes are duplicated here as well.**
3. This PR does the following: Implements filtering to 4 tabs in new contributor admin dashboard. [Mocks](https://www.figma.com/file/bmyOkQwosePcoNAr45Rdk0/Admin-Dashboard?mode=dev).
4. End to end tests for filtering.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct
[Screencast from 09-10-23 04:53:46 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/edf0b50f-e431-4658-9c13-755036a4e6cd)


<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
implemented translation coordinator role and handler changes

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix mppy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc_M2.1

kshitij01042002 committed:
first pass by Sean

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
reviewer changes

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
Sean's comments addressed

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
fixed docstring

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
first draft pass

kshitij01042002 committed:
second draft

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
fixed e2e ci error

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
reviewers commit

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 committed:
merge from develop

kshitij01042002 committed:
first pass

kshitij01042002 committed:
resolved merge conflicts

kshitij01042002 committed:
e2e tests

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed test coverage

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
initial pass

kshitij01042002 committed:
implemented topic filter

kshitij01042002 committed:
fixed frontend tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into gsoc_M2.3

kshitij01042002 committed:
fixed merge conflicts

kshitij01042002 committed:
fixed merge conflicts and minor changes

kshitij01042002 committed:
fixed e2e error

kshitij01042002 committed:
covered backend test coverage

kshitij01042002 committed:
Update gae_models_test.py

kshitij01042002 committed:
fixed failing testsz

kshitij01042002 committed:
Merge branch 'gsoc_M2.3' of https://github.com/kshitij01042002/oppia into gsoc_M2.3

kshitij01042002 committed:
fixed test coverafe

kshitij01042002 committed:
Merge remote-tracking branch 'origin/gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
temp commit

kshitij01042002 committed:
shifted services from suggestion to user

kshitij01042002 committed:
removed unused imports

kshitij01042002 committed:
fixed import errors

kshitij01042002 committed:
fixed UI bugs

kshitij01042002 committed:
minor nit

kshitij01042002 committed:
moved coordinator domain object from suggestion to user domain

kshitij01042002 committed:
fixed type errors

kshitij01042002 committed:
fixed minor errors and implemented in mobile view

kshitij01042002 committed:
added e2e tests

kshitij01042002 committed:
Merge remote-tracking branch 'origin/gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
fixed frontend test coverage

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed ts type checks

kshitij01042002 committed:
fixed ts type checks

kshitij01042002 committed:
fixed ts type checks

kshitij01042002 committed:
covered frontend test coverage

kshitij01042002 committed:
fixed e2e tests

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 commented:
Hi @sagangwee @vojtechjelinek @JayVivarekar @DubeySandeep @heyimShivam can you PTAL? Thanks. I request you to please review as soon as it is possible for you. I am running behind on my GSoC schedule so a quick review would be highly appreciated. Thanks.

kshitij01042002 committed:
review comments by Stephen

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

kshitij01042002 committed:
removed unused vars

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.html:
```diff
@@ -1,67 +1,184 @@
+          </div>
+          <div class="topic-selector-container">
+            <oppia-topic-filter
+              *ngIf="(activeTab === TAB_NAME_TRANSLATION_SUBMITTER) || (activeTab === TAB_NAME_QUESTION_SUBMITTER)"
```
Why don't we have the topic filter for the translation reviewer tab?

kshitij01042002 replied:
Actually all the filters have been planned and implemented in the backend itself. According to the [GSoC Proposal](https://docs.google.com/document/d/19lCEYQUgV7_DwIK_0rz3zslRHX2qKOHn-t9Twpi0qu0/edit#heading=h.vxqaj7i9qj8a), [Admin Dashboard PRD](https://docs.google.com/document/d/1_vvewEwviQjtPvZy9yo74eCcRTQNcQKyCgdHNpyWNWw/edit#heading=h.liiqbmwmz8j8) and [Mocks](https://www.figma.com/file/bmyOkQwosePcoNAr45Rdk0/Admin-Dashboard?node-id=0%3A1&mode=dev).

So we don't have topic filter for reviewers. FYI there is no topic filter for question reviewers as well.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.html:
```diff
@@ -1,67 +1,184 @@
+            </div>
+          </div>
+          <div class="topic-selector-container">
+            <oppia-topic-filter
```
From the screencast, the topic filter is shorter than the last activity filter next to it. Can we make them the same height to be consistent?

kshitij01042002 replied:
The topic filters height changes according to the number of topic selected, so it's not fixed but dynamic. 

Hence I have kept the height to change according to the content in it. Does that sounds reasonable?

sagangwee replied:
Well at the very least I think the heights should be the same in the defaults states—otherwise it looks strange.

kshitij01042002 replied:
I tried fixing it but since both uses different native elements and one's height being dynamically changing I am not able to make it of same height.

sagangwee replied:
The height changing after selecting multiple topics is fine, just change the default heights to match. Should just be some CSS.

kshitij01042002 replied:
I have changed the video in description of the PR. PTAL. Does that looks good?

sagangwee replied:
Hm the filter looks the same to me?

kshitij01042002 replied:
Sorry the video didn't get updated last time. PTAL again.

qinghaoyang replied:
@kshitij01042002 Did you update the video? The filter looks the same to me.

kshitij01042002 replied:
Yeah I have updated it. The topic filter is slightly above but it's the same height now.

qinghaoyang replied:
@sagangwee Does the topic filter look good for you?

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,83 @@
 import { UserService } from 'services/user.service';
 
+
+interface LangaugeChoice {
```
```suggestion
interface LanguageChoice {
```
Typo.

kshitij01042002 replied:
Fixed.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,83 @@
+  allTopicNames: string[] = [];
+  selectedTopicsIds: string[] = [];
+
+  languages: LangaugeChoice[] = [];
```
```suggestion
  languageChoices: LangaugeChoice[] = [];
```

kshitij01042002 replied:
Done.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +134,131 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+              this.assignedLanguageIds = response;
+              this.languages = this.languages.filter((
+                  languageItem) =>
+                this.assignedLanguageIds.includes(languageItem.id));
```
Is this the only place `assignedLanguageIds` is used? If so, then let's get rid of it as a class instance variable.

kshitij01042002 replied:
removed the variable as a whole, we can directly use response instead.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,83 @@
+  loadingMessage!: string;
+
+  selectedLanguage!: string;
+  selectedLanguageId!: string;
```
Instead of creating separate variables for these, can we just have one `selectedLanguageChoice`? Ditto for `selectedTopics`.

kshitij01042002 replied:
Done. PTAL.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,83 @@
+  selectedTopicsIds: string[] = [];
+
+  languages: LangaugeChoice[] = [];
+  lastActivty: number[] = [];
```
```suggestion
  lastActivity: number[] = [];
```
Typo.

kshitij01042002 replied:
Fixed. Thanks for noticing.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +134,131 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+    this.selectedTopicsIds = this.selectedTopics.map((
+        selectedTopic) => {
+      const matchingTopic = this.topics.find(
+        (topicChoice) => topicChoice.topic === selectedTopic);
```
```suggestion
        topicChoice => topicChoice.topic === selectedTopic);
```
Nit: Don't think the parentheses are necessary if there's only one param. Ditto elsewhere.

kshitij01042002 replied:
Fixed. everywhere.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +134,131 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+        selectedTopic) => {
+      const matchingTopic = this.topics.find(
+        (topicChoice) => topicChoice.topic === selectedTopic);
+      return matchingTopic ? matchingTopic.id : '';
```
When would this occur? This feels like an error.

kshitij01042002 replied:
Ideally this would never occur cause the function is called only when there's change in `selectedTopic` via topic filter in frontend. Added this condition to fix ts type errors cause `find` returns undefined as default value.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +134,131 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+      this.filter = new ContributorAdminDashboardFilter(
+        [],
+        this.selectedLanguageId,
+        null);
```
Why doesn't the filter correctly handle the 0 case?

kshitij01042002 replied:
0 means there's no filter at all. In frontend it's shown as `All time`. So when the user selects 'All time' last activity filter is removed.

sagangwee replied:
Okay well then can we have `ContributorAdminDashboardFilter` correctly handle that logic instead of doing that here? "0" seems like a valid input to the filter, so the filter should handle that correctly.

kshitij01042002 replied:
Done. Actually we didn't needed it at all cause the case where value of lastActivity is 0 is already handled in service file.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +134,131 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+    this.translationReviewersCount = this.translationReviewersCountByLanguage[
+      this.selectedLanguageId];
+    this.filter = new ContributorAdminDashboardFilter(
+      [],
```
Does this mean if you select a language, the topic filters are reset? Is that as intended?

kshitij01042002 replied:
No the filters remains same, fixed it.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.ts:
```diff
@@ -184,19 +207,30 @@ export class ContributorDashboardAdminStatsBackendApiService {
+    this.params = {
+      page_size: pageSize,
+      offset: nextOffset,
+      ...(
```
Why is this needed? IIUC this is to copy the object?

kshitij01042002 replied:
This is to pass params only when the filter is selected in frontend other wise don't send it in the get request.

sagangwee replied:
Aren't you already doing with the ternary operators? I'm not sure how the elipses `...` help with that.

kshitij01042002 replied:
Yeah working with ternary operators solves the issue for language and topics but for last activity we have keep `...` cause if there's no last_activity filter selected we are not passing it to the handler.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.ts:
```diff
@@ -298,6 +317,25 @@ export class ContributorDashboardAdminStatsBackendApiService {
   }
+
+  async fetchTopicsData(): Promise<TopicChoice[]> {
+    let classroomDataUrl = this.urlInterpolationService.interpolateUrl(
+      ClassroomDomainConstants.CLASSROOOM_DATA_URL_TEMPLATE, {
+        classroom_url_fragment: 'math'
+      });
+
+    const response = await this.http
+      .get<ClassroomDataBackendDict>(classroomDataUrl)
+      .toPromise();
+
+    const topicChoices: TopicChoice[] = response.topic_summary_dicts.map(
+      (obj) => ({
+        id: obj.id,
+        topic: obj.name
+      }));
+
+    return topicChoices;
```
```suggestion
    const classroomDataUrl = this.urlInterpolationService.interpolateUrl(
      ClassroomDomainConstants.CLASSROOOM_DATA_URL_TEMPLATE, {
        classroom_url_fragment: 'math'
      });
    const response = await this.http
      .get<ClassroomDataBackendDict>(classroomDataUrl)
      .toPromise();
    const topicChoices: TopicChoice[] = response.topic_summary_dicts.map(
      topicSummary => ({
        id: topicSummary.id,
        topic: topicSummary.name
      }));
    return topicChoices;
```

kshitij01042002 replied:
Fixed.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.ts:
```diff
@@ -298,6 +317,25 @@ export class ContributorDashboardAdminStatsBackendApiService {
     });
   }
+
+  async fetchTopicsData(): Promise<TopicChoice[]> {
```
```suggestion
  async fetchTopicChoices(): Promise<TopicChoice[]> {
```

kshitij01042002 replied:
Updated.

sagangwee reviewed (changes requested):
@kshitij01042002 Took a pass on codeowner files, thanks.

kshitij01042002 committed:
review comments by vojtech

kshitij01042002 committed:
review comments by sandeeo

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 committed:
comments bu sagang

kshitij01042002 committed:
fixed lint errors

kshitij01042002 commented:
Hi @sagangwee I have addressed all your comments. PTAL again. Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 committed:
resolved merge conflicts

kshitij01042002 committed:
fixed merge conflicts

kshitij01042002 committed:
Merge branch 'gsoc_M2.4' of https://github.com/kshitij01042002/oppia into gsoc_M2.4

kshitij01042002 committed:
pass 2 by vojtech

kshitij01042002 committed:
wrong commit

kshitij01042002 committed:
Merge branch 'gsoc_M2.4' of https://github.com/kshitij01042002/oppia into gsoc_M2.4

kshitij01042002 committed:
pass 2 by vojtech

kshitij01042002 committed:
Merge branch 'gsoc_M2.3' of https://github.com/kshitij01042002/oppia into gsoc_M2.3

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,92 @@
+  filter: ContributorAdminDashboardFilter = (
+    ContributorAdminDashboardFilter.createDefault());
+
+  selectedTopics: selectedTopicChoices = {
+    ids: [],
+    topics: []
+  };
```
Don't need a new interface for this—just make this an array of `TopicChoice`s like you did for `languageChoices`.

kshitij01042002 replied:
For topic it is different than of language, cause `TopicChoice` has id and topic, but in we need array of ids and array of topics. Using the same thing will make it confusing and difficult to hanfle.

sagangwee replied:
Gotcha. Then maybe just have two variables: `selectedTopicIds` and `selectedTopicNames`. It doesn't look like they need to necessarily be logically grouped via an interface since you only need the IDs or the names at a time.

kshitij01042002 replied:
Yeah sg. Changed it.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,92 @@
+  loadingMessage!: string;
+  selectedLastActivity!: number;
+  allTopicNames: string[] = [];
+
+  languageChoices: LanguageChoice[] = [];
+  lastActivity: number[] = [];
+  filter: ContributorAdminDashboardFilter = (
+    ContributorAdminDashboardFilter.createDefault());
+
+  selectedTopics: selectedTopicChoices = {
+    ids: [],
+    topics: []
+  };
+
+  selectedLanguage: LanguageChoice = {
+    language: '',
+    id: ''
+  };
+
+  topics: TopicChoice[] = [];
```
```suggestion
  languageChoices: LanguageChoice[] = [];
  lastActivity: number[] = [];
  filter: ContributorAdminDashboardFilter = (
    ContributorAdminDashboardFilter.createDefault());
  selectedTopics: selectedTopicChoices = {
    ids: [],
    topics: []
  };
  selectedLanguage: LanguageChoice = {
    language: '',
    id: ''
  };
  topics: TopicChoice[] = [];
```
Nit: The new lines aren't necessary and may even be confusing since it implies some logical separation when there isn't really one.

kshitij01042002 replied:
If we don;t add blank lines between `selectedTopics` and `selectedLanguage` there's a lint error `Expected blank line between class members.`.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,92 @@
+    topics: []
+  };
+
+  selectedLanguage: LanguageChoice = {
+    language: '',
+    id: ''
+  };
```
Can we just not initialize this here? It doesn't really make sense to have a default value for a language choice like this.

kshitij01042002 replied:
If we don't initialize it here, it throughs an error `cannot read properties of undefined` whereever it is used. As it is initialized by response from a request ts assumes it can be undefined if it's not already initialized.

sagangwee replied:
Then that seems like a code smell. It doesn't really make sense to initialize a `LanguageChoice` with empty strings as the default value. Since it is possible for `selectedLanguage` to not be initialized, we should properly handle that in the code. I think initialize `selectedLanguage` to `null` and make sure the accessors of `this.selectedLanguage` handles that case properly (since the initializing `fetchAssignedLanguageIds` call may fail).

kshitij01042002 replied:
Instead of initiating it to null and checking it whenever we access, what if we initialize with empty string and if fetchAssignedLanguageIds fails we could check for the empty string and throw an error. I have implemented it PTAL.

sagangwee replied:
Well, the problem is that `null` is more clear and more accurately conveys the variable state. Initializing the fields as empty strings compiles, but aren't really valid values. Like what does it mean for an `id` or `language` to be an empty string? If you catch an error from `fetchAssignedLanguageIds` and throw there, does that satisfy the compiler?

kshitij01042002 replied:
Yess. The later condition you mentioned satisifes the compiler and we can throw an error there. I have already implemented this.

qinghaoyang replied:
Is it possible to initialize it to `null` and check for the `null`?

kshitij01042002 replied:
If we initialize it to null we have to check each time we are accessing `selectedLanguage` but if we do in this way we only have to check once.

sagangwee replied:
No, I mean adding a `catch` clause to the `fetchAssignedLanguageIds` promise to throw an error to guarantee that `selectedLanguage` is assigned a value or an error is thrown. Then, add a non-null assertion operator (`!`) to the `selectedLanguage` variable.  e.g. https://github.com/oppia/oppia/blob/develop/core/templates/components/checkpoint-celebration-modal/checkpoint-celebration-modal.component.ts#L53-L56

kshitij01042002 replied:
Even after doing this, we still face the behaviur I mentioned above. That is the compiler gives error of `TypeError: Cannot read properties of undefined (reading 'id')`. In case we encounter the condition you mnetioned I have added the condition to throw the error.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -159,11 +160,94 @@ describe('Contributor dashboard Admin page', () => {
+    tick();
+    fixture.detectChanges();
+    expect(component.selectedLanguage.language).toEqual('English');
+    expect(component.selectedLanguage.id).toEqual('en');
```
```suggestion
    spyOn(userService, 'getUserInfoAsync').and.returnValue(
      Promise.resolve(fullAccessUserInfo));
      
    component.ngOnInit();
    tick();
    fixture.detectChanges();
   
    expect(component.selectedLanguage.language).toEqual('English');
    expect(component.selectedLanguage.id).toEqual('en');
```
Make sure to add new lines to differentiate the test phases (setup, action, assertion). This makes tests easy to scan and understand quickly. Ditto elsewhere.

kshitij01042002 replied:
Done. Fixed everywhere. PTAL. Thanks.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -159,11 +160,94 @@ describe('Contributor dashboard Admin page', () => {
+    expect(component.selectedLastActivity).toEqual(0);
+  }));
+
+  it('should apply topic filter', fakeAsync(() => {
```
Can we add a test using both the topic and language filter at the same time? That would have caught the bug noted here:  https://github.com/oppia/oppia/pull/18960#discussion_r1343143317.

kshitij01042002 replied:
I have added the test just below this one, PTAL.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -212,6 +296,46 @@ describe('Contributor dashboard Admin page', () => {
+      spyOn(component, 'checkMobileView').and.returnValue(false);
+      spyOn(userService, 'getUserInfoAsync').and.returnValue(
+        Promise.resolve(fullAccessUserInfo));
+      let fakeClickAwayEvent = new MouseEvent('click');
```
```suggestion
      const fakeClickAwayEvent = new MouseEvent('click');
```
Nit: Prefer `const` if the variable is not being reassigned. Ditto below.

kshitij01042002 replied:
Done.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +143,157 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+              response.includes(languageItem.id));
+            this.selectedLanguage.language = (
+              this.languageChoices[0].language);
+            this.selectedLanguage.id = this.languageChoices[0].id;
```
Nit: Instead of mutating the `this.selectedLanguage` object, reassign it to a new object in one statement.

kshitij01042002 replied:
Sounds reasonable. thanks.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +143,157 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+      .fetchTopicChoices().then(
+        response => {
+          this.topics = response;
+          response.map((
```
Since we're not using the output of `.map()`, just use `forEach` instead. Or better yet, return `topic.topic` in the map function and assign `this.allTopicNames` the new array directly.

kshitij01042002 replied:
Updated the logic. PTAL.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +143,157 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+      selectedTopic => {
+        const matchingTopic = this.topics.find(
+          topicChoice => topicChoice.topic === selectedTopic);
+        return matchingTopic ? matchingTopic.id : '';
```
If this does occur, then we should probably tell the user with an error message toast, otherwise they will be confused as to why their topic filter is not working as intended.

kshitij01042002 replied:
This won't occur in frontend ever cause the selectedTopics is getting selected from this.topics only.

sagangwee replied:
That may be true for now, but may not be true in the future as code changes all the time. Also, that assumption would not be immediately obvious to a future reader less familiar with this code, so at least throw an exception here to self-document the expected behavior.

kshitij01042002 replied:
The point I missed is even if this situation does occur we are passing an empty array to the filter.

sagangwee replied:
You mean an empty string? Yes, but the user selected a topic name in the filter, but we would pass an empty string for that selected topic name, so the filter would do nothing for that selected topic name. That seems like an error.

kshitij01042002 replied:
Yeah that sounds reasonable. Added a condition to throw an error when there is a selected topic but doesn't match any from the list. PTAL.

qinghaoyang replied:
Confirmed

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +143,157 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+      language: 'English',
+      id: 'en'
+    };
+    this.selectedLanguage.language = currentOption.language;
+    this.selectedLanguage.id = currentOption.id;
```
Same here just assign `this.selectedLanguage` in one line.

kshitij01042002 replied:
Sg. Fixed.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +143,157 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+  @HostListener('document:click', ['$event'])
+  onDocumentClick(event: MouseEvent): void {
+    const targetElement = event.target as HTMLElement;
+    if (!this.checkMobileView()) {
```
Flip this conditional into a guard clause to reduce nesting
```suggestion
    if (this.checkMobileView()) {
      return;
    }
    ...
```

kshitij01042002 replied:
Done. Thanks.

sagangwee reviewed (changes requested):
@kshitij01042002 Thanks took another pass.

kshitij01042002 committed:
Merge remote-tracking branch 'origin/gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
pass 2 by Sagang

kshitij01042002 committed:
added back the get method

kshitij01042002 commented:
Hi @sagangwee I have addressed all your comments. PTAL. Thanks for the review.

kshitij01042002 committed:
Merge remote-tracking branch 'origin/gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
added period nit

kshitij01042002 committed:
Merge branch 'gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
comments by Stephen and vojtech

JayVivarekar commented on core/controllers/admin.py:
```diff
@@ -2079,3 +2090,88 @@ def put(self) -> None:
+            },
+            'language_id': {
+                'schema': {
+                    'type': 'basestring'
```
Can we provide validation for `language_id`, so that the handler is safe?

kshitij01042002 replied:
Yeah, actually I was supposed to do that but forgot. Thanks for pointing out.

JayVivarekar reviewed (changes requested):
Thanks @kshitij01042002! Left a comment, PTAL.

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
added validator for languages

kshitij01042002 commented:
Hi @JayVivarekar I have addressed your comments. PTAL again. Thanks for the review.

kshitij01042002 committed:
resolved merge conflicts

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy tests

kshitij01042002 committed:
Merge remote-tracking branch 'origin/gsoc_M2.3' into gsoc_M2.4

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -229,13 +229,43 @@ var AdminPage = function() {
       'Language selector modal taking too long to appear');
     await action.select(
-      'Language selector', languageSelector, 'English'
+      'Language selector', languageSelector, language
```
Where is this language defined?

kshitij01042002 replied:
It's a duplicate function. Fixed it.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -229,13 +229,43 @@ var AdminPage = function() {
+    var languageSelector = $('.e2e-test-language-selector');
+    var languageSelectorCloseButton = $(
+      '.e2e-test-language-selector-close-button');
+    var languageSelectorAddButton = $('.e2e-test-language-selector-add-button');
```
Define these on the top*

kshitij01042002 replied:
Done

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -229,13 +229,43 @@ var AdminPage = function() {
     await action.click('Close button', languageSelectorCloseButton);
     await waitFor.invisibilityOf(
       languageSelectorModal,
       'Language selector modal taking too long to disappear');
```
Can we use _selectLaanuagage here?

kshitij01042002 replied:
It's a duplicate function. Fixed it.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -229,13 +229,43 @@ var AdminPage = function() {
+      'Language selector modal taking too long to disappear');
+  };
+
+  this.makeUserTranslationCoordinator = async function(name, language) {
```
This looks like a duplicate function (line 221)

kshitij01042002 replied:
Yeah, thanks for noiticing! Fixed it.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -247,6 +277,18 @@ var AdminPage = function() {
+  this.addLanguageToCoordinator = async function(name, language) {
+    await this._editUserRole(name);
+
+    var editLanguageButton = $('.e2e-test-edit-language-button');
```
Define this on top*

kshitij01042002 replied:
Done.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -247,6 +277,18 @@ var AdminPage = function() {
+      'Edit coordinated languages button',
+      editLanguageButton);
+
+    await this._selectLanguage(language);
```
Do we need to confirm whether it's added?

kshitij01042002 replied:
I'm not quite sure how can we do that. Cause we are opening a modal selecting a language and then again closing the model so we don't have anything in the frontend after calling the function.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -247,6 +277,18 @@ var AdminPage = function() {
       removeButtonElement, 'Role removal button takes too long to appear.');
   };
 
+  this.addLanguageToCoordinator = async function(name, language) {
```
How is this different from `makeUserTranslationCoordinator`? In case they are similar, can we do something to avoid duplicate pattern/code?

kshitij01042002 replied:
Done. PTAL again. Thanks.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -184,6 +188,18 @@ var ContributorDashboardAdminPage = function() {
   };
 
+  this.switchLanguage = async function(language) {
+    await action.click('Language Selector', languageSelector);
+    await waitFor.visibilityOf(
+      languageDropdown, 'Language Dropdown not visible');
+    var selectorOption = languageSelector.$(
+      `.e2e-test-language-selector-option=${language}`);
+    await action.click(`${language} option selector`, selectorOption);
+    await action.click('Language Selector', languageSelector);
+    await waitFor.invisibilityOf(
+      languageDropdown, 'Language Dropdown did not disappear');
```
USe action.matSelect*

kshitij01042002 replied:
We can't use `action.select/action.matselect` here this is the same situation we discussed in the other PR.[ Reference](https://github.com/oppia/oppia/pull/18951/files/394ac59329e1651784703689bc8577703cc4ce2d#r1342832961). We can only use action.select when we have implemented `<select>` in the code. If we are handling dropdown with a div we can't use it.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -225,9 +241,18 @@ var ContributorDashboardAdminPage = function() {
-    await waitFor.visibilityOf(statsTable, 'stats table is not visible');
-    var statsListItems = await statsListItemsSelector();
-    expect(statsListItems.length).toBe(elementsCount);
+    if (elementsCount === 0) {
```
Instead of adding logic to this block create two functions `expectNoStatsElement`, `expectStatsElementCountToBe`

kshitij01042002 replied:
Done,  PTAL.

DubeySandeep commented on core/tests/webdriverio_utils/ContributorDashboardAdminPage.js:
```diff
@@ -238,6 +263,11 @@ var ContributorDashboardAdminPage = function() {
     await waitFor.invisibilityOf(expandedRow, 'Expanded row still visible');
   };
+
+  this.pageToFullyInitiate = async function() {
```
SHould we chanaeg the name to make it clear what this function is doing else it creates confusion on what it's doing? (I'm thinking of waitForLoadingMessageToDisappear or something around that?)

kshitij01042002 replied:
Yeah, that sounds about right. updated.

DubeySandeep reviewed (changes requested):
@kshitij01042002 Thanks for the PR, I took a pass on e2e test files as codeowner and left a few comments, PTAL!

(The overall changes in e2e test looks great :) )

kshitij01042002 committed:
review comments by Sandeep

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,91 @@
+  topic: string;
+}
+
+export interface selectedTopicChoices {
```
```suggestion
export interface SelectedTopicChoices {
```

kshitij01042002 replied:
Removed it. As per other [comment](https://github.com/oppia/oppia/pull/18960/files/60bad73648ee5dc46d46265b6f9aee45c9bb08f4#r1344512127)

sagangwee replied:
Still there?

kshitij01042002 replied:
Removed!

qinghaoyang replied:
Confirmed.

kshitij01042002 commented:
Hi @DubeySandeep I have addressed your comments. PTAL. Thanks.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/topic-filter/topic-filter.component.ts:
```diff
@@ -0,0 +1,104 @@
+  // and we need to do non-null assertion. For more information, see
+  // https://github.com/oppia/oppia/wiki/Guide-on-defining-types#ts-7-1
+  @Input() listOfDefaultTopics!: string[];
+  @Input() selectedTopics: string[] = [];
```
```suggestion
  @Input() defaultTopicNames!: string[];
  @Input() selectedTopicNames: string[] = [];
```
Nit: To differentiate between topic IDs and names.

kshitij01042002 replied:
Done!

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -84,30 +142,144 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+  }
+
+  applyTopicFilter(): void {
+    this.selectedTopics.ids = this.selectedTopics.topics.map(
```
Instead of allowing the topic-filter component to mutate the `this.selectedTopics.topics` array directly, just assign the new emitted array. In general, try to minimize the use of mutable objects as it's error-prone and can be difficult to trace through logically.

kshitij01042002 replied:
`this.selectedTopics.topics` is used in two way binding to keep track of selected topics continuously and update the stats accordingly.

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/topic-filter/topic-filter.component.ts:
```diff
@@ -0,0 +1,104 @@
+
+  filter(value: string): string[] {
+    const filterValue = value.toLowerCase();
+
```
```suggestion
```

kshitij01042002 replied:
Done

sagangwee reviewed (changes requested):
@kshitij01042002 Thanks, it's looking better. Just a few comments.

JayVivarekar reviewed (approved):
Thanks @kshitij01042002! LGTM for codeowner files.

kshitij01042002 committed:
review comments by Sagang

kshitij01042002 committed:
lint errors fixed

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 commented:
Hi @sagangwee I have addressed all your comments. PTAL. Thanks.

kshitij01042002 committed:
added the null condition

kshitij01042002 committed:
changed logic a but

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
fixed frontend tests

kshitij01042002 committed:
Merge branch 'gsoc_M2.4' of https://github.com/kshitij01042002/oppia into gsoc_M2.4

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM

vojtechjelinek reviewed (approved):
LGTM!

sagangwee reviewed (changes requested):
@kshitij01042002 Just a few outstanding comments. Reminder I will be out until Oct 11th, so I'll leave it up to the other reviewers to approve (doesn't look like you need my explicit approval anyway).

kshitij01042002 committed:
fixed exception errors

kshitij01042002 committed:
removed f from describle

kshitij01042002 commented:
Hi @sagangwee I have addressed all your comments. PTAL whenever you get time.

Also I think I might require your approval for getting the PR merged as oppia/contributor-experience-team codeowner.

qinghaoyang reviewed (commented):
@kshitij01042002 PTAL

kshitij01042002 committed:
fixed minor css

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4

sagangwee commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -17,40 +17,83 @@
+  topics: TopicChoice[] = [];
+  filter: ContributorAdminDashboardFilter = (
+    ContributorAdminDashboardFilter.createDefault());
+
```
```suggestion
```

kshitij01042002 replied:
If I do this there's a lint error `Expected blank line between class members.`

sagangwee reviewed (approved):
@kshitij01042002 LGTM, but please address remaining comments.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.4
